src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,276 @@ PhaseStatus Compiler::optUnrollLoops()
 #pragma warning(pop)
 #endif
 
+//-----------------------------------------------------------------------------
+// optIfConvert
+//
+// Find blocks representing simple if statements represented by conditional jumps
+// over another block. Try to replace the jumps by use of conditional nodes.
+//
+// Arguments:
+//   block -- block that may represent the conditional jump in an if statement.
+//
+// Returns:
+//   true if any IR changes possibly made.
+//
+bool Compiler::optIfConvert(BasicBlock* block)
+{
+#ifndef TARGET_ARM64
+    return false;
+#else
+
+    // Don't optimise the block if it is inside a loop
+    // When inside a loop, branches are quicker than selects.
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->OperIs(GT_JTRUE));
+    if (!last->AsOp()->gtOp1->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->AsOp()->gtOp1->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
+
+    BasicBlock* middleBlock = block;
+    GenTree*    asgNode     = nullptr;
+    Statement*  asgStmt     = nullptr;
+    bool        foundSelect = false;
+
+    // Check the block is followed by a block or chain of blocks that only contain NOPs and
+    // a single ASG statement. The destination of the final block must point to the same as the
+    // true path of the JTRUE block.
+    bool foundMiddle = false;
+    while (!foundMiddle)
+    {
+        middleBlock = middleBlock->bbNext;
+        noway_assert(middleBlock != nullptr);
+
+        if (middleBlock->bbNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            foundMiddle = true;
+        }
+
+        // Make sure there is only one block which jumps to the middle block,
+        // and the middle block is not the start of a TRY block or an exception handler.
+        noway_assert(!fgCheapPredsValid);
+        if (middleBlock->NumSucc() != 1 || middleBlock->bbJumpKind != BBJ_NONE ||
+            middleBlock->bbPreds->flNext != nullptr || middleBlock->bbCatchTyp != BBCT_NONE ||
+            bbIsTryBeg(middleBlock) || bbIsHandlerBeg(middleBlock) ||
+            (middleBlock->bbFlags & BBF_DONT_REMOVE) != 0)
+        {
+            return false;
+        }
+
+        // Can all the nodes within the middle block be made to conditionally execute?
+        for (Statement* const stmt : middleBlock->Statements())
```

Comment:
```suggestion
        for (Statement* const stmt : middleBlock->NonPhiStatements())
```

@AndyAyersMS This should be fine, right?

Reply:
On second thought we shouldn't see any phis here since it has a unique predecessor.